File: tests/benchmark/ackermann/ocaml/ackermann.ml
(* Ackermann function (higher-order definition) *)

let rec iter f n x =
  if n = 0 then
    x
  else
    f (iter f (n - 1) x)

let ackermann m =
  iter (fun f n -> iter f (n + 1) 1) m ((+) 1);;

print_int (ackermann 3 11);;
print_newline ();;

File: tests/benchmark/ackermann/haskell/ackermann.hs
-- Ackermann function (higher-order definition)

compose :: (a -> b) -> (b -> c) -> a -> c
compose f g x = g (f x)

iter :: (a -> a) -> Int -> a -> a
iter f 0 = id
iter f n = compose f (iter f (n - 1))

ackermann :: Int -> Int -> Int
ackermann m =
  iter (\f n -> iter f (n + 1) 1) m (+ 1)

main :: IO ()
main = putStrLn (show (ackermann 3 11))
